pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+		type OracleProvider: ValueProvider<
+			(Self::AccountId, Self::CollectionId),
+			Self::OracleKey,
+			Value = Self::OracleValue,
+			Timestamp = Self::Moment,
+		>;
+
+		/// Used to verify collection admin permissions
+		type IsAdmin: PreConditions<(Self::AccountId, Self::CollectionId), Result = bool>;
+
+		/// Max size of a data collection
+		#[pallet::constant]
+		type MaxCollectionSize: Get<u32>;
+
+		/// Max number of collections
+		#[pallet::constant]
+		type MaxFeedersPerKey: Get<u32>;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type Collection<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		BoundedBTreeMap<T::OracleKey, (T::OracleValue, T::Moment), T::MaxCollectionSize>,
+		ValueQuery,
+	>;
+
+	/// Store the keys that are registed for this collection
+	/// Only keys registered in this stored can be used to create the collection
+	#[pallet::storage]
+	pub(crate) type Keys<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		Blake2_128Concat,
+		T::OracleKey,
+		KeyInfo<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AddedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		RemovedKey {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+		},
+		UpdatedFeeders {
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		},
+		UpdatedCollection {
+			collection_id: T::CollectionId,
+			keys_updated: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account who trigger the action is not a collection admin.
+		IsNotAdmin,
+
+		/// The key is not in the collection.
+		KeyNotInCollection,
+
+		/// Collection size reached
+		MaxCollectionSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update the feeders associated to an specific key.
+		///
+		/// NOTE: In the future this should be go though change guard.
+		/// By now we trust in the admin to choose honest feeders without
+		/// invertor supervision
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(2)]
+		pub fn update_feeders(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				T::IsAdmin::check((who, collection_id)),
+				Error::<T>::IsNotAdmin
+			);
+
+			Self::mutate_and_remove_if_clean(&collection_id, &key, |info| {
+				info.feeders = feeders.clone();
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::UpdatedFeeders {
+				collection_id: collection_id,
+				key: key,
+				feeders,
+			});
+
+			Ok(())
+		}
+
+		/// Update the collection, doing the aggregation for each key in the
+		/// process
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(3)]
+		pub fn update_collection(
```

Comment:
The cost of this method currently is `O(Feeders * Keys)` which could be really high. Just highlight this to be aware of

Reply:
As long as the max weights we set do not consume more than half a block and are still in the realms of enough keys, feeders, this should be fine for now.